File: Codility/6_sorting/number_of_disc_intersections.cpp
// https://codility.com/programmers/task/number_of_disc_intersections/

#include <algorithm>

int solution(vector<int> &A) {
    // y=0 axes intersection x-coordinate and "right side" flag pairs
    vector<pair<int64_t, bool>> discLimits;
    for (size_t i = 0; i < A.size(); ++i)
    {
        discLimits.push_back(make_pair(i - A[i], false));
        discLimits.push_back(make_pair(i + A[i], true));
    }
    
    sort(discLimits.begin(), discLimits.end());
    
    int intersections = 0;
    int openDiscs = 0;
    for (pair<int64_t, bool> limit : discLimits)
    {
        if (limit.second)
        {
            --openDiscs;
            intersections += openDiscs;
            if (intersections > 1e7)
                return -1;
        }
        else
        {
            ++openDiscs;
        }
    }
    
    return intersections;
}

File: Codility/6_sorting/number_of_disc_intersections.py
# https://codility.com/programmers/task/number_of_disc_intersections/

def solution(A):
    disc_limits = []
    for i in xrange(len(A)):
        disc_limits += [(i - A[i], False)]
        disc_limits += [(i + A[i], True)]
        
    disc_limits = sorted(disc_limits)
        
    open_discs = 0
    intersections = 0
    for disc_limit in disc_limits:
        if disc_limit[1]:
            open_discs -= 1
            intersections += open_discs
            if intersections > 1e7:
                return -1
        else:
            open_discs += 1
    
    return intersections
